### PR TITLE
feat(x509): add PEM certificate loading and parsing functionality

### PR DESCRIPTION
**Pull request by Christiantyemele**
closes https://github.com/adorsys/didcomm-mediator-rs/issues/353

Description:
This ticket is to satisfy the valiation rules as from the [section](https://www.ietf.org/archive/id/draft-ietf-oauth-status-list-09.html#name-x509-certificate-extensions) of the spec were the issuer and status issuer are separate entities.

for further reading
https://chatgpt.com/share/67dae3e7-452c-800c-a1ad-1a73920aa908

### CODE REVIEW

File: src/utils/x509_parser.rs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+}
+
+mod test {
+
+    #[test]
+    fn test() {
+        let cert = "./src/test_resources/test_cert.pem";
+        let dec = crate::utils::x509_parser::load_certificate_der(cert).ok();
+        assert!(dec.is_some())
+    }
```
Comment (Hermann-Core) [inline src/utils/x509_parser.rs:32]: Can you add more test suites?
For example, test cases for invalid certificates, already DER-encoded certificates, unsupported certificate formats, and other possible edge cases.

**Review by Hermann-Core — CHANGES_REQUESTED** [review]
I added some comments. Could you check them?

File: Cargo.toml [hunk 2]
```diff
@@ -23,3 +24,4 @@ tower = "0.5"
 tower-http = { version = "0.6", features = ["cors", "trace", "catch-panic"] }
 tracing = "0.1"
 tracing-subscriber = "0.3"
+x509-parser = "0.17"
```
[the remark this answers is not in the document]
Comment (Christiantyemele) [reply Cargo.toml:27]: :+1:

File: src/utils/x509_parser.rs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+/// Loads a certificate from a PEM file and returns the DER bytes
+#[inline]
+pub fn load_certificate_der<P: AsRef<Path>>(cert_path: P) -> Result<Vec<u8>> {
+    // Read the PEM-encoded certificate file
+    let cert_pem = fs::read_to_string(&cert_path)
+        .with_context(|| format!("Failed to read certificate from {:?}", cert_path.as_ref()))?;
+
+    // Parse the PEM
+    let pem = parse(cert_pem).context("Failed to parse PEM format")?;
```
[the remark this answers is not in the document]
Comment (Christiantyemele) [reply src/utils/x509_parser.rs:14]: i think anyhow can handle this errors, since they are mostly fs errors :stroke:

File: src/utils/x509_parser.rs [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        }
+    }
+    #[test]
+    fn test_der_encoded_certificate_content() {
+        // Pretend this is DER content (binary-like junk data in a .pem)
+        let test_cert = "./src/test_resources/test_cert.der";
+        let res = crate::utils::x509_parser::load_certificate_der(test_cert).ok();
+        assert!(res.is_none());
+    }
+}
```
Comment (ndefokou) [inline src/utils/x509_parser.rs:54]: I think you can add some other test cases

- A file with a `.pem` extension but invalid content.
- A file with a different extension but valid PEM content.
:smiling_face_with_tear:

File: src/utils/x509_parser.rs [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+/// Loads a certificate from a PEM file and returns the DER bytes
+#[inline]
+pub fn load_certificate_der<P: AsRef<Path>>(cert_path: P) -> Result<Vec<u8>, Error> {
+    let path = cert_path.as_ref();
+
+    // Check file extension is .pem
+    if path.extension().and_then(|s| s.to_str()).unwrap_or("") != "pem" {
+        Err(Error::InvalidFileType)?;
+    }
```
Comment (ndefokou) [inline src/utils/x509_parser.rs:15]: Instead of using unwrap_or(""), you could directly check if the extension is Some("pem") what do you think 

e.g
if path.extension().and_then(|s| s.to_str()) != Some("pem") {
    return Err(Error::InvalidFileType);
}

File: src/utils/x509_parser.rs (hunk 4 above)
Comment (Christiantyemele) [reply src/utils/x509_parser.rs:54]: okay there is no harm in doing that

File: src/utils/x509_parser.rs (hunk 5 above)
Comment (Christiantyemele) [reply src/utils/x509_parser.rs:15]: okay

**Review by ndefokou — APPROVED** [review]
Good

**Review by Blindspot22 — APPROVED** [review]
Looks good :rocket:
